---
title: "Statistical analysis of contingency tables with infer"
tags: [infer]
categories: [statistical analysis]
type: learn-subsection
---

<script src="/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<p>This article requires that you have the following packages installed: modeldata and tidymodels.</p>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<div id="introduction-1" class="section level3">
<h3>Introduction</h3>
<p>In this vignette, we’ll walk through conducting a <span class="math inline">\(\chi^2\)</span> (chi-squared) test of independence and a chi-squared goodness of fit test using <code>infer</code>. We’ll start out with a chi-squared test of independence, which can be used to test the association between two categorical variables. Then, we’ll move on to a chi-squared goodness of fit test, which tests how well the distribution of one categorical variable can be approximated by some theoretical distribution.</p>
<p>Throughout this vignette, we’ll make use of the <code>ad_data</code> dataset supplied by the <code>modedata</code> package, which contains data related to cognitive impairment in 333 patients from <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3079734/">Craig-Schapiro <em>at al</em> (2011)</a>. See <code>?ad_data</code> for more information on the variables included and their source. One of the main research questions in these data were how does a person’s genetics related to the Apolipoprotein E gene affect their cognitive skills (as had been reported in the literature). The data shows:</p>
<pre class="r"><code>library(tidymodels) # Includes the infer package

data(ad_data, package = &quot;modeldata&quot;)</code></pre>
<p>The three main genetic variants are called E2, E3, and E4. The values above represent the genetic makeup of patients based on what they inherited from their parents (i.e, a value of “E2E4” means E2 from one parent and E4 from the other).</p>
</div>
<div id="test-of-independence" class="section level3">
<h3>Test of Independence</h3>
<p>To carry out a chi-squared test of independence, we’ll examine the association between their cognitive ability (impaired and healthy) and the genetic makeup. This is what the relationship looks like in the sample data:</p>
<p><img src="figs/plot-indep-1.svg" width="672" /></p>
<p>If there were no relationship, we would expect to see the purple bars reaching to the same height, regardless of cognitive ability. Are the differences we see here, though, just due to random noise?</p>
<p>First, to calculate the observed statistic, we can use <code>specify()</code> and <code>calculate()</code>.</p>
<pre class="r"><code># calculate the observed statistic
observed_indep_statistic &lt;- ad_data %&gt;%
  specify(Genotype ~ Class) %&gt;%
  calculate(stat = &quot;Chisq&quot;)</code></pre>
<p>The observed <span class="math inline">\(\chi^2\)</span> statistic is 21.5774809. Now, we want to compare this statistic to a null distribution, generated under the assumption that these variables are not actually related, to get a sense of how likely it would be for us to see this observed statistic if there were actually no association between cognitive ability and genetics.</p>
<p>We can <code>generate</code> the null distribution in one of two ways—using randomization or theory-based methods. The randomization approach permutes the response and explanatory variables, so that each person’s genetics is matched up with a random cognitive rating from the sample in order to break up any association between the two.</p>
<pre class="r"><code># generate the null distribution using randomization
null_distribution_simulated &lt;- ad_data %&gt;%
  specify(Genotype ~ Class) %&gt;%
  hypothesize(null = &quot;independence&quot;) %&gt;%
  generate(reps = 5000, type = &quot;permute&quot;) %&gt;%
  calculate(stat = &quot;Chisq&quot;)</code></pre>
<p>Note that, in the line <code>specify(Genotype ~ Class)</code> above, we could use the equivalent syntax <code>specify(response = Genotype, explanatory = Class)</code>. The same goes in the code below, which generates the null distribution using theory-based methods instead of randomization.</p>
<pre class="r"><code># generate the null distribution by theoretical approximation
null_distribution_theoretical &lt;- ad_data %&gt;%
  specify(Genotype ~ Class) %&gt;%
  hypothesize(null = &quot;independence&quot;) %&gt;%
  # note that we skip the generation step here!
  calculate(stat = &quot;Chisq&quot;)</code></pre>
<p>To get a sense for what these distributions look like, and where our observed statistic falls, we can use <code>visualize()</code>:</p>
<pre class="r"><code># visualize the null distribution and test statistic!
null_distribution_simulated %&gt;%
  visualize() + 
  shade_p_value(observed_indep_statistic,
                direction = &quot;greater&quot;)</code></pre>
<p><img src="figs/visualize-indep-1.svg" width="672" /></p>
<p>We could also visualize the observed statistic against the theoretical null distribution. Note that we skip the <code>generate()</code> and <code>calculate()</code> steps when using the theoretical approach, and that we now need to provide <code>method = "theoretical"</code> to <code>visualize()</code>.</p>
<pre class="r"><code># visualize the theoretical null distribution and test statistic!
ad_data %&gt;%
  specify(Genotype ~ Class) %&gt;%
  hypothesize(null = &quot;independence&quot;) %&gt;%
  visualize(method = &quot;theoretical&quot;) + 
  shade_p_value(observed_indep_statistic,
                direction = &quot;greater&quot;)</code></pre>
<p><img src="figs/visualize-indep-theor-1.svg" width="672" /></p>
<p>To visualize both the randomization-based and theoretical null distributions to get a sense of how the two relate, we can pipe the randomization-based null distribution into <code>visualize()</code>, and further provide <code>method = "both"</code>.</p>
<pre class="r"><code># visualize both null distributions and the test statistic!
null_distribution_simulated %&gt;%
  visualize(method = &quot;both&quot;) + 
  shade_p_value(observed_indep_statistic,
                direction = &quot;greater&quot;)</code></pre>
<p><img src="figs/visualize-indep-both-1.svg" width="672" /></p>
<p>Either way, it looks like our observed test statistic would be fairly unlikely if there were actually no association between cognition and genotype More exactly, we can calculate the p-value:</p>
<pre class="r"><code># calculate the p value from the observed statistic and null distribution
p_value_independence &lt;- null_distribution_simulated %&gt;%
  get_p_value(obs_stat = observed_indep_statistic,
              direction = &quot;greater&quot;)

p_value_independence
#&gt; # A tibble: 1 x 1
#&gt;   p_value
#&gt;     &lt;dbl&gt;
#&gt; 1  0.0002</code></pre>
<p>Thus, if there were really no relationship between cognition and genotype, the probability that we would see a statistic as or more extreme than 21.5774809 is approximately 2^{-4}.</p>
<p>Note that, equivalently to the steps shown above, the package supplies a wrapper function, <code>chisq_test</code>, to carry out Chi-Squared tests of independence on tidy data. The syntax goes like this:</p>
<pre class="r"><code>chisq_test(ad_data, Genotype ~ Class)
#&gt; # A tibble: 1 x 3
#&gt;   statistic chisq_df  p_value
#&gt;       &lt;dbl&gt;    &lt;int&gt;    &lt;dbl&gt;
#&gt; 1      21.6        5 0.000630</code></pre>
</div>
<div id="goodness-of-fit" class="section level3">
<h3>Goodness of Fit</h3>
<p>Now, moving on to a chi-squared goodness of fit test, we’ll take a look at just the genotype data. Many papers have investigated the relationship of Apolipoprotein E to diseases. For example, <a href="https://annals.org/aim/article-abstract/717641/meta-analysis-apolipoprotein-e-genotypes-risk-coronary-heart-disease">Song <em>et al</em> (2004)</a> conducted a meta-analysis of numerous studies that looked at this gene and heart disease. In their paper, they describe the frequency of the different genotypes across many samples. For the cognition study, it might be interesting to see if our sample of genotypes was consistent with this literature (treating the rates, for this analysis, as known).</p>
<p>The rates of the meta-analysis and our observed data are:</p>
<pre class="r"><code># Song, Y., Stampfer, M. J., &amp; Liu, S. (2004). Meta-Analysis: Apolipoprotein E 
# Genotypes and Risk for Coronary Heart Disease. Annals of Internal Medicine, 
# 141(2), 137.
meta_rates &lt;- c(&quot;E2E2&quot; = 0.71, &quot;E2E3&quot; = 11.4, &quot;E2E4&quot; = 2.32,
                &quot;E3E3&quot; = 61.0, &quot;E3E4&quot; = 22.6, &quot;E4E4&quot; = 2.22)
meta_rates &lt;- meta_rates/sum(meta_rates) # these add up to slightly &gt; 100%

obs_rates &lt;- table(ad_data$Genotype)/nrow(ad_data)
round(cbind(obs_rates, meta_rates) * 100, 2)
#&gt;      obs_rates meta_rates
#&gt; E2E2      0.60       0.71
#&gt; E2E3     11.11      11.37
#&gt; E2E4      2.40       2.31
#&gt; E3E3     50.15      60.85
#&gt; E3E4     31.83      22.54
#&gt; E4E4      3.90       2.21</code></pre>
<p>Suppose our null hypothesis is that <code>Genotype</code> follows the same frequency distribution as the meta-analysis. Lets now test whether this difference in distributions is statistically significant.</p>
<p>First, to carry out this hypothesis test, we would calculate our observed statistic.</p>
<pre class="r"><code># calculating the null distribution
observed_gof_statistic &lt;- ad_data %&gt;%
  specify(response = Genotype) %&gt;%
  hypothesize(null = &quot;point&quot;, p = meta_rates) %&gt;%
  calculate(stat = &quot;Chisq&quot;)</code></pre>
<p>The observed statistic is 23.3838483. Now, generating a null distribution, by just dropping in a call to <code>generate()</code>:</p>
<pre class="r"><code># generating a null distribution
null_distribution_gof &lt;- ad_data %&gt;%
  specify(response = Genotype) %&gt;%
  hypothesize(null = &quot;point&quot;, p = meta_rates) %&gt;%
  generate(reps = 5000, type = &quot;simulate&quot;) %&gt;%
  calculate(stat = &quot;Chisq&quot;)</code></pre>
<p>Again, to get a sense for what these distributions look like, and where our observed statistic falls, we can use <code>visualize()</code>:</p>
<pre class="r"><code># visualize the null distribution and test statistic!
null_distribution_gof %&gt;%
  visualize() + 
  shade_p_value(observed_gof_statistic,
                direction = &quot;greater&quot;)</code></pre>
<p><img src="figs/visualize-indep-gof-1.svg" width="672" /></p>
<p>This statistic seems like it would be unlikely if our rates were the same as the rates from the meta-analysis! How unlikely, though? Calculating the p-value:</p>
<pre class="r"><code># calculate the p-value
p_value_gof &lt;- null_distribution_gof %&gt;%
  get_p_value(observed_gof_statistic,
              direction = &quot;greater&quot;)

p_value_gof
#&gt; # A tibble: 1 x 1
#&gt;   p_value
#&gt;     &lt;dbl&gt;
#&gt; 1  0.0002</code></pre>
<p>Thus, if each genotype occurred at the same rate as the Song paper, the probability that we would see a distribution like the one we did is approximately 2^{-4}.</p>
<p>Again, equivalently to the steps shown above, the package supplies a wrapper function, <code>chisq_test</code>, to carry out Chi-Squared goodness of fit tests on tidy data. The syntax goes like this:</p>
<pre class="r"><code>chisq_test(ad_data, response = Genotype, p = meta_rates)
#&gt; # A tibble: 1 x 3
#&gt;   statistic chisq_df  p_value
#&gt;       &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
#&gt; 1      23.4        5 0.000285</code></pre>
</div>
</div>
<div id="session-information" class="section level1">
<h1>Session information</h1>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────────
#&gt;  setting  value                       
#&gt;  version  R version 3.6.1 (2019-07-05)
#&gt;  os       macOS Mojave 10.14.6        
#&gt;  system   x86_64, darwin15.6.0        
#&gt;  ui       X11                         
#&gt;  language (EN)                        
#&gt;  collate  en_US.UTF-8                 
#&gt;  ctype    en_US.UTF-8                 
#&gt;  tz       America/New_York            
#&gt;  date     2020-03-26                  
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────
#&gt;  package    * version    date       lib source                             
#&gt;  broom      * 0.5.4      2020-01-27 [1] CRAN (R 3.6.0)                     
#&gt;  dials      * 0.0.4.9000 2020-03-21 [1] local                              
#&gt;  dplyr      * 0.8.5      2020-03-07 [1] CRAN (R 3.6.0)                     
#&gt;  ggplot2    * 3.3.0      2020-03-05 [1] CRAN (R 3.6.0)                     
#&gt;  infer      * 0.5.1      2019-11-19 [1] CRAN (R 3.6.0)                     
#&gt;  parsnip    * 0.0.5.9000 2020-03-26 [1] local                              
#&gt;  purrr      * 0.3.3      2019-10-18 [1] CRAN (R 3.6.0)                     
#&gt;  recipes    * 0.1.10     2020-03-18 [1] CRAN (R 3.6.1)                     
#&gt;  rlang        0.4.5.9000 2020-03-21 [1] Github (r-lib/rlang@a90b04b)       
#&gt;  rsample    * 0.0.5.9000 2020-03-23 [1] Github (tidymodels/rsample@4fdbd6c)
#&gt;  tibble     * 2.1.3      2019-06-06 [1] CRAN (R 3.6.0)                     
#&gt;  tidymodels * 0.1.0      2020-02-16 [1] CRAN (R 3.6.0)                     
#&gt;  tune       * 0.0.1.9000 2020-03-21 [1] Github (tidymodels/tune@6694258)   
#&gt;  workflows  * 0.1.0      2019-12-30 [1] CRAN (R 3.6.1)                     
#&gt;  yardstick  * 0.0.5      2020-01-23 [1] CRAN (R 3.6.0)                     
#&gt; 
#&gt; [1] /Library/Frameworks/R.framework/Versions/3.6/Resources/library</code></pre>
</div>
